File: codigo.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <list>
#include <chrono>
#include <string>

using namespace std;

struct Record {
    string key;
    string value; // Guardaremos el Score de felicidad como string
};

// --- TABLA HASH: ENCADENAMIENTO SEPARADO ---
class HashTableChaining {
private:
    int size;
    vector<list<Record>> table;
    int collisions;

public:
    HashTableChaining(int tableSize) : size(tableSize), collisions(0) {
        table.resize(size);
    }

    int hashFunction(string key) {
        unsigned long hash = 0;
        for (char c : key) hash = hash * 31 + c;
        return hash % size;
    }

    void insert(string key, string value) {
        int index = hashFunction(key);
        if (!table[index].empty()) collisions++;
        table[index].push_back({key, value});
    }

    string search(string key) {
        int index = hashFunction(key);
        for (auto& r : table[index]) {
            if (r.key == key) return r.value;
        }
        return "No encontrado";
    }

    int getCollisions() { return collisions; }
    double loadFactor(int total) { return (double)total / size; }
};

// --- TABLA HASH: SONDEO LINEAL ---
class HashTableLinear {
private:
    int size;
    vector<Record> table;
    vector<bool> occupied;
    int collisions;

public:
    HashTableLinear(int tableSize) : size(tableSize), collisions(0) {
        table.resize(size);
        occupied.resize(size, false);
    }

    int hashFunction(string key) {
        unsigned long hash = 0;
        for (char c : key) hash = hash * 31 + c;
        return hash % size;
    }

    void insert(string key, string value) {
        int index = hashFunction(key);
        while (occupied[index]) {
            collisions++;
            index = (index + 1) % size;
        }
        table[index] = {key, value};
        occupied[index] = true;
    }

    string search(string key) {
        int index = hashFunction(key);
        int start = index;
        while (occupied[index]) {
            if (table[index].key == key) return table[index].value;
            index = (index + 1) % size;
            if (index == start) break;
        }
        return "No encontrado";
    }

    int getCollisions() { return collisions; }
};

// Función para leer el CSV 2019
vector<pair<string, string>> readHappinessCSV(string filename) {
    vector<pair<string, string>> data;
    ifstream file(filename);
    string line, country, score, temp;

    getline(file, line); // Saltar cabecera
    while (getline(file, line)) {
        stringstream ss(line);
        getline(ss, temp, ',');    // Overall rank
        getline(ss, country, ','); // Country or region
        getline(ss, score, ',');   // Score
        data.push_back({country, score});
    }
    return data;
}

int main() {
    string filename = "2019.csv";
    int tableSize = 311;
    auto data = readHappinessCSV(filename);

    HashTableChaining hashChain(tableSize);
    HashTableLinear hashLinear(tableSize);

    // --- PRUEBA ENCADENAMIENTO ---
    auto start = chrono::high_resolution_clock::now();
    for (auto& p : data) hashChain.insert(p.first, p.second);
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> t_ins_chain = end - start;

    // --- PRUEBA SONDEO LINEAL ---
    start = chrono::high_resolution_clock::now();
    for (auto& p : data) hashLinear.insert(p.first, p.second);
    end = chrono::high_resolution_clock::now();
    chrono::duration<double> t_ins_linear = end - start;

    // --- SALIDA DE RESULTADOS ---
    cout << "INFORME PRACTICA 3 - C++ (Dataset 2019)" << endl;
    cout << "========================================" << endl;
    cout << "METODO: ENCADENAMIENTO" << endl;
    cout << "Tiempo Insercion: " << t_ins_chain.count() << "s" << endl;
    cout << "Colisiones: " << hashChain.getCollisions() << endl;
    cout << "Factor de Carga: " << hashChain.loadFactor(data.size()) << endl << endl;

    cout << "METODO: SONDEO LINEAL" << endl;
    cout << "Tiempo Insercion: " << t_ins_linear.count() << "s" << endl;
    cout << "Colisiones: " << hashLinear.getCollisions() << endl;

    return 0;
}

File: codigo.py
import pandas as pd
import time
import matplotlib.pyplot as plt

# --- ESTRUCTURAS DE DATOS MANUALES ---

class HashTableChaining:
    def __init__(self, size):
        self.size = size
        self.table = [[] for _ in range(size)]
        self.collisions = 0

    def hash_function(self, key):
        # Usamos el hash nativo de Python y lo ajustamos al tamaño de la tabla
        return abs(hash(str(key))) % self.size

    def insert(self, key, value):
        index = self.hash_function(key)
        if len(self.table[index]) > 0:
            self.collisions += 1
        self.table[index].append((key, value))

    def search(self, key):
        index = self.hash_function(key)
        for k, v in self.table[index]:
            if k == key:
                return v
        return None

    def load_factor(self, total_elements):
        return total_elements / self.size

class HashTableLinearProbing:
    def __init__(self, size):
        self.size = size
        self.table = [None] * size
        self.collisions = 0

    def hash_function(self, key):
        return abs(hash(str(key))) % self.size

    def insert(self, key, value):
        index = self.hash_function(key)
        while self.table[index] is not None:
            self.collisions += 1
            index = (index + 1) % self.size
            # Evitar bucle infinito si la tabla se llena
        self.table[index] = (key, value)

    def search(self, key):
        index = self.hash_function(key)
        start = index
        while self.table[index] is not None:
            if self.table[index][0] == key:
                return self.table[index][1]
            index = (index + 1) % self.size
            if index == start:
                break
        return None

    def load_factor(self, total_elements):
        return total_elements / self.size

# --- PROCESAMIENTO DE DATOS (2019.csv) ---

# Cargar dataset
try:
    df = pd.read_csv("2019.csv")
    df = df.dropna()
    # Claves: Nombres de países, Valores: Puntuación de felicidad
    keys = df["Country or region"].astype(str).tolist()
    data_records = df.to_dict('records')
    
    table_size = 311  # Número primo para reducir colisiones
    
    # 1. Prueba con Encadenamiento Separado
    hash_chain = HashTableChaining(table_size)
    start_time = time.time()
    for i, key in enumerate(keys):
        hash_chain.insert(key, data_records[i])
    insert_time_chain = time.time() - start_time

    start_time = time.time()
    for key in keys:
        hash_chain.search(key)
    search_time_chain = time.time() - start_time

    # 2. Prueba con Sondeo Lineal
    hash_linear = HashTableLinearProbing(table_size)
    start_time = time.time()
    for i, key in enumerate(keys):
        hash_linear.insert(key, data_records[i])
    insert_time_linear = time.time() - start_time

    start_time = time.time()
    for key in keys:
        hash_linear.search(key)
    search_time_linear = time.time() - start_time

    # 3. Diccionario Nativo de Python
    native_dict = {}
    start_time = time.time()
    for i, key in enumerate(keys):
        native_dict[key] = data_records[i]
    insert_time_dict = time.time() - start_time

    start_time = time.time()
    for key in keys:
        native_dict.get(key)
    search_time_dict = time.time() - start_time

    # --- REPORTE Y GRÁFICOS ---

    results = pd.DataFrame({
        "Metodo": ["Encadenamiento", "Sondeo lineal", "Diccionario Python"],
        "Tiempo insercion": [insert_time_chain, insert_time_linear, insert_time_dict],
        "Tiempo busqueda": [search_time_chain, search_time_linear, search_time_dict],
        "Colisiones": [hash_chain.collisions, hash_linear.collisions, 0],
        "Factor de carga": [hash_chain.load_factor(len(keys)), hash_linear.load_factor(len(keys)), len(keys)/table_size]
    })

    print("\n--- RESULTADOS DE LA PRÁCTICA 3 (Dataset 2019) ---")
    print(results)

    # Gráfico Comparativo
    fig, ax = plt.subplots(1, 2, figsize=(12, 5))

    ax[0].bar(results["Metodo"], results["Tiempo insercion"], color='skyblue')
    ax[0].set_title("Tiempo de Inserción (s)")
    ax[0].set_ylabel("Segundos")

    ax[1].bar(results["Metodo"], results["Tiempo busqueda"], color='salmon')
    ax[1].set_title("Tiempo de Búsqueda (s)")
    ax[1].set_ylabel("Segundos")

    plt.tight_layout()
    plt.show()

except FileNotFoundError:
    print("Error: No se encontró el archivo '2019.csv'.")
